ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -0,0 +1,243 @@
+    operationTracker = new SimpleOperationTracker(routerConfig.routerDatacenterName, blobId.getPartition(), true,
+        routerConfig.routerGetSuccessTarget, routerConfig.routerGetRequestParallelism);
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with a getBlobInfo operation.
+   * @return {@link MessageFormatFlags#BlobInfo}
+   */
+  @Override
+  MessageFormatFlags getOperationFlag() {
+    return MessageFormatFlags.BlobInfo;
+  }
+
+  /**
+   * For this operation, create and populate get requests (in the form of {@link RequestInfo}) to send out.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  @Override
+  void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
+    //First, check if any of the existing requests have timed out.
+    Iterator<Map.Entry<Integer, GetRequestInfo>> inFlightRequestsIterator =
+        correlationIdToGetRequestInfo.entrySet().iterator();
+    while (inFlightRequestsIterator.hasNext()) {
+      Map.Entry<Integer, GetRequestInfo> entry = inFlightRequestsIterator.next();
+      if (time.milliseconds() - entry.getValue().startTimeMs > routerConfig.routerRequestTimeoutMs) {
+        operationTracker.onResponse(entry.getValue().replicaId, false);
+        setOperationException(
+            new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut));
+        inFlightRequestsIterator.remove();
+      } else {
+        // the entries are ordered by correlation id and time. Break on the first request that has not timed out.
+        break;
+      }
+    }
+
+    checkAndMaybeComplete();
+    if (isOperationComplete()) {
+      return;
+    }
+
+    Iterator<ReplicaId> replicaIterator = operationTracker.getReplicaIterator();
+    while (replicaIterator.hasNext()) {
+      ReplicaId replicaId = replicaIterator.next();
+      String hostname = replicaId.getDataNodeId().getHostname();
+      Port port = new Port(replicaId.getDataNodeId().getPort(), PortType.PLAINTEXT);
+      GetRequest getRequest = createGetRequest(blobId, getOperationFlag());
+      RequestInfo request = new RequestInfo(hostname, port, getRequest);
+      int correlationId = getRequest.getCorrelationId();
+      correlationIdToGetRequestInfo.put(correlationId, new GetRequestInfo(replicaId, time.milliseconds()));
+      requestRegistrationCallback.registerRequestToSend(this, request);
+      replicaIterator.remove();
+    }
+  }
+
+  /**
+   * Handle the body of the response: Deserialize and set the {@link BlobInfo} to return.
+   * @param payload the body of the response.
+   * @throws IOException if there is an IOException while deserializing the body.
+   * @throws MessageFormatException if there is a MessageFormatException while deserializing the body.
+   */
+  void handleBody(InputStream payload)
```

Comment:
just a habit from the good old C days, where a method has to be defined before it is used (much like variables). Anyway, moved it below and made it private.